public/js/components/Scopes.js

```diff
@@ -3,110 +3,169 @@
-const { DOM: dom } = React;
-const Accordion = React.createFactory(require("./Accordion"));
+const { getPause, getLoadedObjects } = require("../selectors");
+const { DOM: dom, PropTypes } = React;
+const ManagedTree = React.createFactory(require("./util/ManagedTree"));
+const Arrow = React.createFactory(require("./util/Arrow"));
 
-require("./Scopes.css");
+function info(text) {
+  return dom.div({ className: "pane-info" }, text);
+}
 
 function getScopes(pauseInfo) {
   if (!pauseInfo) {
-    return [];
+    return null;
   }
 
-  const environment = pauseInfo.getIn(["frame", "environment"]);
-  let scope = environment;
-  const scopes = [ scope ];
-
-  while (scope = scope.get("parent")) { // eslint-disable-line no-cond-assign
-    scopes.push(scope);
+  let scope = pauseInfo.getIn(["frame", "scope"]);
+  if (!scope) {
+    return null;
   }
 
-  return scopes;
-}
-
-function renderArgumentProperty(argumentBinding) {
-  const argumentName = argumentBinding.keySeq().first();
-  const actor = argumentBinding.first().getIn(["value", "actor"]);
-  return dom.li(
-    { className: "scope-property-list-item", key: actor },
-    argumentName
-  );
-}
-
-function renderVariableProperty(variableBinding, variableName) {
-  const actor = variableBinding.getIn(["value", "actor"]);
-  return dom.li(
-    { className: "scope-property-list-item", key: actor },
-    variableName
-  );
-}
-
-function renderScopeBindings(bindings) {
-  const variableBindings = bindings.get("variables");
-  const argumentBindings = bindings.get("arguments");
-
-  return dom.ul(
-    { className: "scope-property-list" },
-    argumentBindings.map(renderArgumentProperty),
-    variableBindings.map(renderVariableProperty)
-  );
-}
-
-function renderFunctionScope(scope, isFirst) {
-  const displayName = scope.getIn(["function", "displayName"]);
-  const functionName = displayName || "(anonymous)";
-  const label = "Function";
-
-  return {
-    header: `${label} [${functionName}]`,
-    component: () => renderScopeBindings(scope.get("bindings")),
-    opened: isFirst
-  };
-}
-
-/* TODO: render block with variables */
-function renderBlockScope(scope, isFirst) {
-  const label = "Block Scope";
-  return {
-    header: label,
-    component: () => renderScopeBindings(scope.get("bindings")),
-    opened: isFirst
-  };
-}
-
-function renderObjectScope(scope, isFirst) {
-  const objectClass = scope.getIn(["object", "class"]);
-  const label = objectClass;
+  const scopes = [];
+
+  do {
+    const type = scope.get("type");
+
+    if (type === "function" || type === "block") {
+      const bindings = scope.get("bindings");
+      const vars = bindings
+        .get("arguments").map(arg => arg.entrySeq().get(0))
+        .concat(bindings.get("variables").entrySeq())
+        .filter(binding => (!binding[1].hasIn(["value", "missingArguments"]) &&
+                            !binding[1].hasIn(["value", "optimizedOut"])))
+        .sort((a, b) => a[0].localeCompare(b[0]))
```

Comment:
Gets all the variables in scope, filters out ones that are optimized out, and sorts them.